### ISSUE TITLE
RFC: Updating Ghost-CLI to use Yarn rather than NPM

### ISSUE THREAD

**Issue by acburdine** [TryGhost/Ghost-CLI#112]
In an effort to speed up install times and reduce issues/errors with npm install (especially where memory usage is considered), I did some research into potentially moving to Yarn instead, and here are the results of that research/experimentation.

~~The first issue was that yarn doesn't output json correctly on `yarn info`, but that has been fixed as of yarn 0.18.~~

The biggest issue/hurdle to overcome is that yarn does not take into account the yarn.lock file of dependencies that it's installing. This makes sense from yarn's perspective, as to do so would mean inspecting the contents of the npm package before downloading, which would be a huge pain. However, this means that any attempt to install Ghost directly would not use the yarn.lock and just install whichever dependencies it finds initially.

The second issue is that a direct yarn install installs each dependency into the root `node_modules` folder and then links them into all of the subfolders. While this doesn't necessarily prevent things from working, it does make things a bit more complex considering the current method of installing ghost via npm and then copying it into the correct version folder.

To address both these issues, I propose this solution:

The install process would go back to kind of what it was before the switch to npm-installed ghost, and instead of using npm to download the built Ghost file, the download capability would be built directly into the CLI. The download itself could come from npm/yarn itself (the packaged tar.gz file) or Github (the built zip). Personally I'd recommend the npm package, as it's in the same location each time and doesn't require messing around with Github's api. Also, I think yarn has an ability to show the location of a npm package download based on the version and package name? (needs more research).

Once the ghost package is downloaded/placed in the right version folder, then `yarn install --production` would be run inside of the downloaded Ghost folder.

So, in summary:

### Pros:
- faster install times (in theory)
- less memory usage (especially when compared to npm@3)
- bypass `npm shrinkwrap` idiosyncracies between npm major versions

### Cons:
- Makes the CLI install process a little more complex
- installing yarn via npm is currently deprecated (I think? needs more research)

### TODOs (if this is to be done)
- Add the yarn.lock file to Ghost itself (and ensure they are included in the release build)
- Re-build install/update processes
- all the testing

### More needed research:
- yarn programmatic API (does it exist? any way to get install process completion percentages?)
- yarn install avenues (npm?)
- does `yarn global add ghost-cli` use the lockfile (if one were to be included with ghost-cli)?

**Comment by acburdine** [comment]
After a little more research - here's how I think the install process could work with Yarn:

1. `yarn info ghost@<version> dist --json` - returns both a shasum and download link for the tar package
2. Download tar and verify shasum integrity using [this library](https://www.npmjs.com/package/shasum)
3. Untargz using whatever method is necessary
4. run `yarn` inside the appropriate directory.

**Comment by acburdine** [comment]
So this may have to be pushed back to later - yarn seems to have issues when using the `--production` flag 😕

If we end up adding a yarn.lock to Ghost itself then I can do further research, but for the moment it seems that without a `yarn.lock` file, `yarn install --production` decides to install all the dependencies, which is decidedly less than ideal.

**Comment by ErisDS** [comment]
What is the issue with adding a yarn.lock file to Ghost? Is it just the we-tend-not-to-add-too-many-things-to-root rule? If so, I'm all in favour of waiving that for yarn.

I am also very, very keen to separate the download, and install steps in Ghost-CLI. If I want to install Ghost 3 times, I shouldn't have to download the zip/tar etc 3 times?

**Comment by acburdine** [comment]
After discussion, the plan is to move ahead with using Yarn as the installation mechanism in Ghost-CLI.

**Comment by ErisDS** [comment]
Now that this is progressing, do you have any numbers on speed and memory usage?

E.g. Is there real value in switching? The issue lists it as maybe 😉

**Comment by acburdine** [comment]
It definitely seems to be faster - when it's not trying to install devDependencies. As far as memory usage I haven't analyzed it yet with the CLI but it's definitely better as I've seen it with other projects.

I'll do some memory profiling once I've figured out why yarn keeps insisting on installing dev dependencies

**Comment by acburdine** [comment]
@ErisDS so here's an initial investigation into time + memory usage. This is disregarding the CLI overhead, which should be fairly minimal.

# Environment
- Ghost 1.0.0-alpha.10 downloaded from npm
- Memory process observed with Activity Monitor
- Time used observed with `time`

# Yarn
**Command: `yarn install --production --no-emoji --no-progress`**
Output of `time`: `227.59 real        36.45 user        20.43 sys`.
Max memory usage I observed: < 200 mb - fairly consistent

# NPM (version 2.15 - didn't try v3)
**Command: `npm install --production`**
Output of `time`: `92.71 real        75.03 user        34.25 sys`
Max memory usage observed: > 600mb - increased over time from about 300mb

# Conclusions:
In the end, I think that Yarn is probably the better solution - esp. as these numbers are on `install` - and I bet the time would decrease by a lot dependent on cache & on update of Ghost b/c yarn wouldn't be linking _everything_. Also, the fact that npm (v2) uses a third of the ram that npm does is something to take into account.

**Comment by acburdine** [comment]
More updates (after further testing on an ubuntu vm) - Yarn install completes in about 26s, whereas npm takes > 70.
